fastqpreprocessing/src/htslib_tagsort.cpp

```diff
@@ -326,41 +298,158 @@ void write_out_partial_txt_file(std::vector<TAGTUPLE> const& tuple_records,
             << get<14>(tuple_records[record_index]) /* record[13] */ << "\n";
   }
 
-  outfile.close(); // TODO is closing specifically here relevant to the mutex?
-  g_mtx.lock();
-  partial_files->push_back(tempfile);
-  g_mtx.unlock();
+  return tempfile_filename;
 }
 
-unsigned int g_num_thread_deallocations = 0;
-
-void process_alignments(INPUT_OPTIONS_TAGSORT& options, bam1_t** alns,
-                        bam_hdr_t* bamHdr, unsigned int buf_no, unsigned int n,
-                        std::vector<std::string>* partial_files)
+// Manages worker threads' access to reading the input file. Any worker can take
+// any line from the file, but only one thread can be reading at a time. This
+// class lets them take turns: they call readAlignments() whenever they want
+// more data, and it blocks until they can have it.
+class AlignmentReader
 {
-  std::vector<TAGTUPLE> tuple_records;
-  // TODO this maps from a string to a heap-allocated copy of that string.
-  std::unordered_map<std::string, std::string*> string_map;
+public:
+  explicit AlignmentReader(INPUT_OPTIONS_TAGSORT options) : options_(options)
+  {
+    if ((sam_file_ptr_ = hts_open(options.bam_input.c_str(),"r")) == nullptr)
+      crash(options.bam_input + ": cannot open file.");
 
-  for (unsigned int i = 0; i < n; i++)
-    parseOneAlignment(&tuple_records, alns[i], options, bamHdr, string_map);
+    bam_hdr_ = sam_hdr_read(sam_file_ptr_); //read header
 
-  write_out_partial_txt_file(tuple_records, options.temp_folder, partial_files);
+    allocateAlignmentBuffers();
+  }
+
+  ~AlignmentReader()
+  {
+    for (unsigned int i = 0; i < options_.nthreads; i++)
+    {
+      for (unsigned int k = 0; k < options_.alignments_per_batch; k++)
+        bam_destroy1(aln_arr_[i][k]);
+      free(aln_arr_[i]);
+    }
+    free(aln_arr_);
+    sam_hdr_destroy(bam_hdr_);
+    hts_close(sam_file_ptr_);
+  }
+
+  // Blocks until it's this thread's turn to read, then reads a batch of alignments.
+  // Returns a pointer to the alignment ptr array, and number of alignment ptrs in the array.
+  std::pair<bam1_t**, unsigned int> readAlignments(int thread_index)
+  {
+    const std::lock_guard<std::mutex> lock(mutex_);
+    unsigned int cur_num_read = 0;
+    while (cur_num_read < options_.alignments_per_batch)
+    {
+      if (sam_read1(sam_file_ptr_, bam_hdr_, aln_arr_[thread_index][cur_num_read]) == 0)
```

Comment:
The existing code treats sam_read1() > 0 as a success, but everything I could find about the library seemed to indicate == 0 means success. We should figure out which is correct.